python/ray/_private/worker.py

```diff
@@ -121,13 +121,17 @@
 DAGNode = TypeVar("DAGNode")
 
 
+class RayWaitable(Generic[R]):
```

Comment:
This could be contraversial.

I found a bit weird to make ObjectRef and ObjectRefGenerator a subclass of RayWaitable because

- they don't share any functionality
- they don't share any interface.

So I found it a bit more natural to just define the Union types. Any thoughts?

Reply:
using a union makes sense. but I think using an empty interface won't hurt either.  whichever way should be fine.

2 related questions:
1. Should `StreamingObjectRefGenerator` also be a generic type? 
2. is there any difference between defining `RayWaitable` in this way versus just `RayWaitable[R] = Union["ObjectRef[R]", StreamingObjectRefGenerator]`?

Reply:
1. so in this case, it is the type of returned object ref? 

StreamingObjectRefGenerator[T] == `__next__(self) -> ObjectRef[T]`?

I think this sgtm.

2. So I tried defining `RayWaitable = Union["ObjectRef[R]", StreamingObjectRefGenerator]`, but it raised an exception, so I tried the current approach lol..

Reply:
`class RayWaitable` and `Union["ObjectRef[R]", StreamingObjectRefGenerator]` are fundamentally different? looks `Union["ObjectRef[R]", StreamingObjectRefGenerator]` are more preferrable

Reply:
Confirmed it doesn't work. We cannot make RayWaittable a Generic type without declaring a class that inherits from Generic